controller/src/main/java/org/jboss/as/controller/persistence/ConfigurationFile.java

```diff
@@ -353,14 +353,18 @@ private File determineMainFile(final String rawName, final String name) {
         if (mainName == null) {
             // Try the basic case, where name is the name
             final File directoryFile = new File(configurationDir, name);
-            if (directoryFile.exists()) {
-                mainName = stripPrefixSuffix(name); // TODO what if the stripped name doesn't exist? And why would there be a file like configuration/standalone.last.xml?
-            } else if (interactionPolicy.isReadOnly()) {
-                // We allow absolute paths in this case
-                final File absoluteFile = new File(name);
-                if (absoluteFile.exists()) {
-                    return absoluteFile;
+            try {
+                if (directoryFile.exists()
+                        && (directoryFile.getCanonicalPath().startsWith(configurationDir.getCanonicalPath()) || interactionPolicy.isReadOnly())) {
+                    mainName = stripPrefixSuffix(name); // TODO what if the stripped name doesn't exist? And why would there be a file like configuration/standalone.last.xml?
+                } else if (interactionPolicy.isReadOnly()) {
+                    // We allow absolute paths in this case
+                    final File absoluteFile = new File(name);
+                    if (absoluteFile.exists()) {
+                        return absoluteFile;
+                    }
                 }
+            } catch (IOException e) {
```

Comment:
Well, probably throw the exception `throw ControllerLogger.ROOT_LOGGER.mainFileNotFound(name != null ? name : rawName, configurationDir);` on this catch would be the safest action. 

Go out of this try..catch with `mainName` equals to null would affect the processing of the following `if`check (notice InteractionPolicy.NEW and InteractionPolicy.DISCARD). If you get an IOException I think going out with an error would be the safest option.

Reply:
Do you think it would affect it? If `mainName == null` because of the IOException I don't expect the following checks to produce anything useful either so we reach the throw at the end of the method and I didn't want to duplicate that line.

Reply:
Hello @michpetrov , sorry for the delay, I missed the notification of your comment.

The IOException is thrown by the explicit check you are adding to detect whether the configuration file is under the configuration directory. I think if the code doesn't react properly when you are unable to detect such a circumstance, you could be bypassing it. At first, glance it looks more secure to not allow it explicitly throwing the exception.

For example, as it is now, you will be returning at [L374](https://github.com/wildfly/wildfly-core/blob/0fe0b252c2cbf12fadf502bf406d5bb3a8be887c/controller/src/main/java/org/jboss/as/controller/persistence/ConfigurationFile.java#L374) without knowing if  `directoryFile.getCanonicalPath().startsWith(configurationDir.getCanonicalPath())` was true or false. 

It seems it only will happen when the iteration policy is DISCARD and NEW, looking at their descriptions, I'm not sure if that is correct.

Could you confirm if iteration policy is DISCARD or NEW you want to allow a configuration file outside of the configuration directory?

Reply:
@yersan I've looked into Java's IO and it's currently impossible for the IOException to be thrown, so while I have to put the try-catch in there having it throw an exception is kinda useless

Java checks for invalid files*, an invalid file would throw the exception however an invalid file wouldn't pass the `directoryFile.exists()` check

So if a user somehow got an invalid file in it would short-circuit before checking the paths; in that case it might indeed return at L374 but if that happened we'd throw an exception at [L211](https://github.com/wildfly/wildfly-core/blob/0fe0b252c2cbf12fadf502bf406d5bb3a8be887c/controller/src/main/java/org/jboss/as/controller/persistence/ConfigurationFile.java#L211)

\* "invalid file" currently means the path contains a null byte

Reply:
Hi @michpetrov , I understand your arguments, but we should be more explicit so when you are reading the code, you can understand the intention. Having an empty IOException catch doesn't look good, however, if I see explicitly an exception thrown there, I can explicitly understand the intention; don't allow a file configuration outside of the configuration directory. Having a duplicate line is not too bad, and once the code is compiled, it will use the most effective path.

If we leave it as it is, in the future, someone reading this piece could go back and forth with the same questions we are discussing here. If I read the JavaDoc of [getCannonicalPath()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/File.html#getCanonicalPath()), it is difficult to arrive at your conclusion of when the IOException is thrown.

Would that convince you?

> So if a user somehow got an invalid file in it would short-circuit before checking the paths; in that case it might indeed return at L374 but if that happened we'd throw an exception at L211

Right, although It makes the code dependent on the calling method, it is a private method but ... Anyway, this looks like another weak case, in the case of `InteractionPolicy.NEW` and `InteractionPolicy.DISCARD`, if the directory doesn't exist, do you want to allow a configuration file outside of the configuration directory?

Reply:
Based on the description I thought it wouldn't have an effect but it looks like DISCARD will empty a file that's outside the config folder, so that's not good. I'm gonna have to refactor this, thanks for pointing that out.